### DISCUSSION TITLE
Expose more options of the gzip/zlib compress/decompress methods

### DISCUSSION THREAD

**Discussion by rvenson** [godotengine/godot-proposals#6348] [Engine Core]
I have some compressed strings in my project read from an external file. The strings are base64 encoded and compressed using zlib (deflate), containing no header/trailer. I use the following example code to get the final data:

input example:
````
"S61IzC3ISVVISSxJBAA="
````

````gdscript
func parse(base_string):
    var raw = Marshalls.base64_to_raw(base_string)
    var data = raw.decompress_dynamic(-1, File.COMPRESSION_DEFLATE)
    var text = data.get_string_from_utf8()
    return text
````

output expected:
````
"example data"
````

but returns the error:
````
decompress_dynamic: incorrect header check
core/io/compression.cpp:244 @ decompress_dynamic()
````

---

As my strings have no header, I discovered that I should use a negative value (-15) for the window bits on the decompress, different from the default (15). So I guess there's no way to decompress successfully with GDscript since there's no option to change this parameter. The only option exposed for zlib in the global config is the compression level.

Here are some references to the window_bits options:
- https://stackoverflow.com/a/29020296
- https://docs.python.org/3/library/zlib.html#zlib.compress

![image](https://user-images.githubusercontent.com/7561122/221053487-15520d75-a8f2-4f7d-90ec-284c460b4cf0.png)
_Current exposed options for compression_

---

Maybe there are other workarounds, I'm not an expert and I may be missing something, but considering that the compression level parameters are already exposed in the project configs, I suppose that the best option is to expose the `window_bits` value too. The implementation should be uncomplicated, actually, I already made it, but just looking for help to make sure that this is a possible PR.

**Comment by AThousandShips** [comment]
I would recommend adding "MODE_DEFLATE_RAW" which just runs -15 as window bits, exposing window bits globally would probably not be very helpful, if you want both raw and non-raw mode you'd have to mess with global settings and I doubt there would be any great need or reason to pick any other window size than 15 or -15.

**Reply by rvenson** [reply]
I agree. Since the only difference between zlib and gzip is the window bits value, makes more sense to have another mode for the raw. The only concern is about the confusing mode name. All three modes use deflate actually. Maybe it's a good idea to rename MODE_DEFLATE to MODE_ZLIB, as was suggested in an [old issue](https://github.com/godotengine/godot/issues/22440#issuecomment-424926641), keeping the MODE_DEFLATE_RAW to the raw mode.

**Reply by AThousandShips** [reply]
Probably best not to cause a compatibility breaking change
